File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/repository/ReviewPromptManager.kt
package com.rork.guidestreamtvandroid.data.repository

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.content.SharedPreferences
import android.util.Log
import com.rork.guidestreamtvandroid.data.local.DeviceSessionService
import com.google.android.play.core.review.ReviewManagerFactory
import java.util.concurrent.TimeUnit

/**
 * In-app review gate — mirrors iOS ReviewPromptManager.swift.
 *
 * Counters are stored per install, NOT per user id: the large majority of
 * installs never sign in, so keying on a user would exclude them entirely.
 *
 * Nothing here ever prompts mid-task. A qualifying action only *arms* the
 * prompt; it is presented on the next return to the app (onResume), which is
 * the moment the user has just come back from a streaming app or finished
 * something and is not being interrupted.
 */
object ReviewPromptManager {

    enum class Trigger(val value: String) {
        DEEP_LINK_RETURN("deeplink_return"),
        WATCHED_MILESTONE("watched_milestone"),
        ALERT_TO_WATCH("alert_to_watch"),
        CAST_STARTED("cast_started"),
        RETURN_VISIT("return_visit"),
    }

    private const val PREFS = "gs_review_prompt"
    private const val KEY_FIRST_LAUNCH = "firstLaunch"
    private const val KEY_SESSIONS = "sessions"
    private const val KEY_DEEP_LINKS = "deepLinks"
    private const val KEY_WATCHED = "watched"
    private const val KEY_SHOWN_DATES = "shownDates"
    private const val KEY_ARMED_DEEP_LINK = "armedDeepLink"
    private const val KEY_ACTIVE_DAYS = "activeDays"
    private const val KEY_LAST_ACTIVE_DAY = "lastActiveDay"

    private const val MIN_INSTALL_DAYS = 3
    private const val MIN_SESSIONS = 3
    private const val DEEP_LINK_THRESHOLD = 2
    private const val WATCHED_THRESHOLD = 5
    /** Separate calendar days the app has been opened, for [Trigger.RETURN_VISIT]. */
    private const val ACTIVE_DAY_THRESHOLD = 3
    // 90 days is the industry floor between prompts, and the annual cap now
    // matches what the platform allows rather than self-throttling below it —
    // over quota the flow silently does nothing, so a held-back prompt is
    // simply a forfeited one. Mirrors iOS.
    private const val COOLDOWN_DAYS = 90
    private const val MAX_PER_YEAR = 3

    private lateinit var prefs: SharedPreferences

    /** Set when a qualifying action fires; consumed by [maybePresent]. */
    @Volatile private var pendingTrigger: Trigger? = null

    fun init(context: Context) {
        prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        if (!prefs.contains(KEY_FIRST_LAUNCH)) {
            prefs.edit().putLong(KEY_FIRST_LAUNCH, System.currentTimeMillis()).apply()
        }
    }

    private val ready: Boolean get() = ::prefs.isInitialized

    // MARK: - Signals

    fun noteSessionStarted() {
        if (!ready) return
        bump(KEY_SESSIONS)
    }

    /** A deep link into a streaming app. Arms, but never prompts right now —
     *  the user is about to leave; the prompt waits for their return. */
    fun noteDeepLinkFired() {
        if (!ready) return
        val count = bump(KEY_DEEP_LINKS)
        if (count >= DEEP_LINK_THRESHOLD) {
            prefs.edit().putBoolean(KEY_ARMED_DEEP_LINK, true).apply()
        }
    }

    fun noteWatchedToggled() {
        if (!ready) return
        val count = bump(KEY_WATCHED)
        if (count >= WATCHED_THRESHOLD && count % WATCHED_THRESHOLD == 0) {
            consider(Trigger.WATCHED_MILESTONE)
        }
    }

    fun noteCastStarted() {
        if (!ready) return
        consider(Trigger.CAST_STARTED)
    }

    fun noteAlertOpened() {
        if (!ready) return
        prefs.edit().putBoolean(KEY_ARMED_DEEP_LINK, true).apply()
    }

    // MARK: - Presentation

    /** Call from Activity.onResume. Presents at most one prompt. */
    fun maybePresent(activity: Activity) {
        if (!ready) return
        noteActiveDay()
        if (prefs.getBoolean(KEY_ARMED_DEEP_LINK, false)) {
            prefs.edit().putBoolean(KEY_ARMED_DEEP_LINK, false).apply()
            consider(Trigger.DEEP_LINK_RETURN)
        } else if (prefs.getInt(KEY_ACTIVE_DAYS, 0) >= ACTIVE_DAY_THRESHOLD) {
            // Coming back on a third separate day IS the satisfaction signal.
            // This app is a guide — its value is delivered by looking
            // something up and leaving — so the prompt is not gated on a deep
            // link or a watched-toggle first. Mirrors iOS.
            consider(Trigger.RETURN_VISIT)
        }
        val trigger = pendingTrigger ?: return
        pendingTrigger = null
        launchFlow(activity, trigger)
    }

    private fun consider(trigger: Trigger) {
        if (!isEligible) return
        pendingTrigger = trigger
    }

    private val isEligible: Boolean
        get() {
            val firstLaunch = prefs.getLong(KEY_FIRST_LAUNCH, System.currentTimeMillis())
            val installDays = TimeUnit.MILLISECONDS.toDays(System.currentTimeMillis() - firstLaunch)
            if (installDays < MIN_INSTALL_DAYS) return false
            if (sessionsForGate < MIN_SESSIONS) return false
            // Never on top of a forced-update or update-nudge sheet.
            if (runCatching { AppUpdateGate.get().prompt }.getOrNull() != null) return false

            val shown = shownDates()
            val now = System.currentTimeMillis()
            val yearAgo = now - TimeUnit.DAYS.toMillis(365)
            if (shown.count { it > yearAgo } >= MAX_PER_YEAR) return false
            val last = shown.maxOrNull() ?: return true
            return TimeUnit.MILLISECONDS.toDays(now - last) >= COOLDOWN_DAYS
        }

    /**
     * The local counter starts at zero on the launch where this shipped, so it
     * under-reports a long-standing install. DeviceSessionService keeps the
     * same per-install number from the day the app was installed — take the
     * larger, so an update does not make a returning user start over.
     */
    private val sessionsForGate: Int
        get() = maxOf(
            prefs.getInt(KEY_SESSIONS, 0),
            runCatching { DeviceSessionService.get().sessionCount }.getOrDefault(0),
        )

    /** One increment per calendar day the app is opened. */
    private fun noteActiveDay() {
        val today = java.text.SimpleDateFormat("yyyy-MM-dd", java.util.Locale.US)
            .format(java.util.Date())
        if (prefs.getString(KEY_LAST_ACTIVE_DAY, null) == today) return
        prefs.edit().putString(KEY_LAST_ACTIVE_DAY, today).apply()
        bump(KEY_ACTIVE_DAYS)
    }

    private fun launchFlow(activity: Activity, trigger: Trigger) {
        val manager = ReviewManagerFactory.create(activity)
        manager.requestReviewFlow().addOnCompleteListener { request ->
            if (!request.isSuccessful) {
                Log.w("GSReview", "requestReviewFlow failed: ${request.exception?.message}")
                return@addOnCompleteListener
            }
            manager.launchReviewFlow(activity, request.result).addOnCompleteListener {
                // Play never tells us whether the sheet was actually shown or
                // whether the user rated. Record the attempt either way — that
                // is what the quota and the cooldown have to be based on.
                recordShown(trigger)
            }
        }
    }

    private fun recordShown(trigger: Trigger) {
        val dates = (shownDates() + System.currentTimeMillis()).takeLast(10)
        prefs.edit().putString(KEY_SHOWN_DATES, dates.joinToString(",")).apply()
        runCatching {
            WatchIntentLogger.get().log(
                eventType = WatchIntentLogger.IntentEventType.REVIEW_PROMPT_REQUESTED,
                metadata = mapOf(
                    "trigger" to trigger.value,
                    "deep_links" to prefs.getInt(KEY_DEEP_LINKS, 0),
                    "watched" to prefs.getInt(KEY_WATCHED, 0),
                    "sessions" to sessionsForGate,
                    "active_days" to prefs.getInt(KEY_ACTIVE_DAYS, 0),
                ),
            )
        }
    }

    // MARK: - Helpers

    private fun shownDates(): List<Long> =
        prefs.getString(KEY_SHOWN_DATES, "")
            .orEmpty()
            .split(",")
            .mapNotNull { it.trim().toLongOrNull() }

    private fun bump(key: String): Int {
        val next = prefs.getInt(key, 0) + 1
        prefs.edit().putInt(key, next).apply()
        return next
    }

    private const val PACKAGE = "com.rork.guidestreamtvandroid"

    /**
     * Opens the Play Store listing for the explicit "Rate" row.
     *
     * Deliberately NOT [launchReviewFlow]: Play's in-app sheet is quota'd and
     * silently does nothing once spent, so a button wired to it would look
     * broken exactly when a user has decided to leave a review. The listing
     * always works.
     *
     * Tries the Play Store app first (`market://`) and falls back to the web
     * listing on devices without it.
     */
    fun openStoreListing(context: Context) {
        val market = Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$PACKAGE"))
            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        try {
            context.startActivity(market)
        } catch (_: ActivityNotFoundException) {
            val web = Intent(
                Intent.ACTION_VIEW,
                Uri.parse("https://play.google.com/store/apps/details?id=$PACKAGE"),
            ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            runCatching { context.startActivity(web) }
        }
    }
}

File: ios/GuideStreamTV/Services/ReviewPromptManager.swift
//
//  ReviewPromptManager.swift
//  GuideStreamTV
//
//  Decides when to ask for an App Store rating, per the plan in
//  claude/in-app-review-prompt-plan-aug30-2026.md.
//
//  Two rules shape everything here:
//
//  1. NO PRE-PROMPT, EVER. Both stores prohibit review gating — no "do you
//     like the app?", no star row, nothing overlaid on the system card. This
//     type never renders UI; it only decides *when* the native sheet should be
//     asked for. If a future change adds a custom question in front of it,
//     that is a policy violation, not a design choice.
//
//  2. COUNTERS ARE PER INSTALL, NOT PER ACCOUNT. 810 of 1,083 devices in the
//     telemetry have never signed in. Keying any of this on user_id would
//     halve the eligible population at every gate, so everything below lives
//     in UserDefaults against the install.
//
//  The API itself is quota-limited and silent: Apple allows at most 3 prompts
//  per user per 365 days, never reports whether the sheet appeared, and never
//  shows it in TestFlight. That is why nothing user-visible may depend on it,
//  and why the manual "Rate GuideStream TV" row opens the store URL instead of
//  calling this.
//

import Foundation
import SwiftUI

@MainActor
@Observable
final class ReviewPromptManager {
    static let shared = ReviewPromptManager()

    /// The success moment that armed the current request. Nil when nothing is
    /// pending. The root view observes this, calls the SwiftUI
    /// `requestReview` action, then calls `didPresent()`.
    private(set) var pendingTrigger: Trigger?

    enum Trigger: String {
        /// Came back to the app after a deep link — the core promise kept.
        case deepLinkReturn = "deeplink_return"
        /// Accumulated habit, no interruption cost.
        case watchedMilestone = "watched_milestone"
        /// A new-episode alert led straight to watching it. Highest peak.
        case alertToWatch = "alert_to_watch"
        /// A cast to Roku / Tizen started. Rarest, kept at the user's request.
        case castStarted = "cast_started"
        /// Came back on a third separate day, having actually used the app.
        case returnVisit = "return_visit"
    }

    // MARK: Thresholds

    private let minInstallDays = 3
    private let minSessions = 3
    private let deepLinkThreshold = 2
    private let watchedThreshold = 5
    /// Separate calendar days the app has been opened, for `.returnVisit`.
    private let activeDayThreshold = 3
    /// 90 days is the industry floor between prompts. The annual cap is now
    /// Apple's own 3 rather than a self-imposed 2: the API silently no-ops
    /// over quota, so holding one back buys no goodwill — it forfeits a
    /// prompt, which at this install volume is a real cost.
    private let cooldownDays = 90
    private let maxPerYear = 3

    // MARK: Storage (per install)

    private let defaults = UserDefaults.standard
    private let firstLaunchKey = "gs.reviewPrompt.firstLaunch"
    private let sessionsKey = "gs.reviewPrompt.sessions"
    private let deepLinksKey = "gs.reviewPrompt.deepLinks"
    private let watchedKey = "gs.reviewPrompt.watched"
    private let shownDatesKey = "gs.reviewPrompt.shownDates"
    private let armedKey = "gs.reviewPrompt.armedDeepLink"
    private let activeDaysKey = "gs.reviewPrompt.activeDays"
    private let lastActiveDayKey = "gs.reviewPrompt.lastActiveDay"

    private init() {
        if defaults.object(forKey: firstLaunchKey) == nil {
            defaults.set(Date(), forKey: firstLaunchKey)
        }
    }

    // MARK: - Signals
    //
    // Fed from WatchIntentLogger.log, which already sees every one of these.
    // Hooking there rather than at each call site means a new success moment
    // cannot be added to the app and silently miss this.

    func noteSessionStarted() { bump(sessionsKey) }

    /// A deep link fired. Deliberately does NOT arm the prompt: the success is
    /// coming *back*, and a sheet presented at tap time would be buried under
    /// the launching app. `appDidBecomeActive()` closes the loop.
    func noteDeepLinkFired() {
        bump(deepLinksKey)
        defaults.set(true, forKey: armedKey)
    }

    func noteWatchedToggled() {
        bump(watchedKey)
        // Every fifth, not every toggle past the fifth — matches Android.
        let count = defaults.integer(forKey: watchedKey)
        if count >= watchedThreshold, count % watchedThreshold == 0 {
            consider(.watchedMilestone)
        }
    }

    func noteCastStarted() { consider(.castStarted) }

    func noteAlertOpened() { defaults.set(true, forKey: armedKey) }

    /// Called when the app returns to the foreground. If a deep link or an
    /// alert took the user out, coming back is the moment the promise was
    /// kept.
    func appDidBecomeActive() {
        noteActiveDay()

        if defaults.bool(forKey: armedKey) {
            defaults.set(false, forKey: armedKey)
            if defaults.integer(forKey: deepLinksKey) >= deepLinkThreshold {
                consider(.deepLinkReturn)
                return
            }
        }

        // Coming back on a third separate day is as strong a satisfaction
        // signal as a run of deep links, and far more people clear it: in the
        // twelve days after this shipped, only 16 iOS devices fired a deep
        // link at all. It still requires one real action on the install, so
        // presence alone never asks.
        guard defaults.integer(forKey: activeDaysKey) >= activeDayThreshold,
              defaults.integer(forKey: deepLinksKey) >= 1
                || defaults.integer(forKey: watchedKey) >= 1
        else { return }
        consider(.returnVisit)
    }

    /// One increment per calendar day the app is opened.
    private func noteActiveDay() {
        let today = Self.dayStamp()
        guard defaults.string(forKey: lastActiveDayKey) != today else { return }
        defaults.set(today, forKey: lastActiveDayKey)
        bump(activeDaysKey)
    }

    private static func dayStamp() -> String {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: Date())
    }

    // MARK: - Gating

    private func consider(_ trigger: Trigger) {
        guard pendingTrigger == nil, isEligible else { return }
        pendingTrigger = trigger
    }

    private var isEligible: Bool {
        // Never in TestFlight: Apple does not show the sheet there, so asking
        // would burn a trigger and teach us nothing.
        if isTestFlight { return false }
        guard installedDays >= minInstallDays else { return false }
        guard sessionsForGate >= minSessions else { return false }
        // Never stack on top of other modal UI.
        guard CoachMarkManager.shared.homeTourDone else { return false }
        guard AppUpdateGate.shared.prompt == nil else { return false }
        // Do not ask for a rating in a session that has already gone wrong.
        guard WatchIntentLogger.shared.recentErrors.isEmpty else { return false }
        let recent = shownDates
        guard recent.count < maxPerYear else { return false }
        if let last = recent.max(),
           Date().timeIntervalSince(last) < Double(cooldownDays) * 86_400 {
            return false
        }
        return true
    }

    /// The local counter starts at zero on the launch where this file first
    /// shipped, so it under-reports someone who has been here for months.
    /// DeviceSessionService keeps the same per-install number from the day the
    /// app was installed — take whichever is larger, so the gate does not make
    /// a returning user re-earn their history because they updated.
    private var sessionsForGate: Int {
        max(defaults.integer(forKey: sessionsKey), DeviceSessionService.shared.sessionCount)
    }

    private var installedDays: Int {
        guard let first = defaults.object(forKey: firstLaunchKey) as? Date else { return 0 }
        return Int(Date().timeIntervalSince(first) / 86_400)
    }

    /// A sandbox receipt means TestFlight (or a simulator build).
    private var isTestFlight: Bool {
        Bundle.main.appStoreReceiptURL?.lastPathComponent == "sandboxReceipt"
    }

    /// Prompt dates inside the last 365 days, pruned on read.
    private var shownDates: [Date] {
        let cutoff = Date().addingTimeInterval(-365 * 86_400)
        return (defaults.array(forKey: shownDatesKey) as? [Date] ?? []).filter { $0 > cutoff }
    }

    // MARK: - Presentation

    /// Called by the root view immediately after invoking `requestReview`.
    /// Neither store reports whether the card appeared or what the user did,
    /// so this event count against the store's rating count is the only
    /// measurement available.
    func didPresent() {
        guard let trigger = pendingTrigger else { return }
        pendingTrigger = nil
        defaults.set(shownDates + [Date()], forKey: shownDatesKey)
        WatchIntentLogger.shared.log(
            eventType: .reviewPromptRequested,
            metadata: ["trigger": trigger.rawValue]
        )
    }

    private func bump(_ key: String) {
        defaults.set(defaults.integer(forKey: key) + 1, forKey: key)
    }

    // MARK: - Store URL (the manual entry point)

    /// The App Store write-review page. The manual "Rate GuideStream TV" row
    /// opens THIS rather than calling `requestReview`: over quota, or in
    /// TestFlight, the API is a silent no-op and the row reads as a broken
    /// button. A user reported exactly that in GS-01008.
    static let writeReviewURL = URL(string: "https://apps.apple.com/app/id6773443577?action=write-review")!
}
